In [8]:
import numpy as np
import pandas as pd

In [10]:
# Load the data
ratings = pd.read_csv('ratings.csv')
movies = pd.read_csv('movies.csv')

In [11]:
ratings

userId  movieId  rating   timestamp
0              1        1     4.0  1225734739
1              1      110     4.0  1225865086
2              1      158     4.0  1225733503
3              1      260     4.5  1225735204
4              1      356     5.0  1225735119
...          ...      ...     ...         ...
33832157  330975     8340     2.0  1091583256
33832158  330975     8493     2.5  1091585709
33832159  330975     8622     4.0  1091581777
33832160  330975     8665     3.0  1091581765
33832161  330975     8783     2.5  1091581799

[33832162 rows x 4 columns]

In [12]:
movies

movieId                                              title  \
0            1                                   Toy Story (1995)   
1            2                                     Jumanji (1995)   
2            3                            Grumpier Old Men (1995)   
3            4                           Waiting to Exhale (1995)   
4            5                 Father of the Bride Part II (1995)   
...        ...                                                ...   
86532   288967               State of Siege: Temple Attack (2021)   
86533   288971                                 Ouija Japan (2021)   
86534   288975   The Men Who Made the Movies: Howard Hawks (1973)   
86535   288977                    Skinford: Death Sentence (2023)   
86536   288983  UNZIPPED: An Autopsy of American Inequality (2...   

                                            genres  
0      Adventure|Animation|Children|Comedy|Fantasy  
1                       Adventure|Children|Fantasy  
2                                   Comedy|Romance  
3                             Comedy|Drama|Romance  
4                                           Comedy  
...                                            ...  
86532                                 Action|Drama  
86533                                Action|Horror  
86534                                  Documentary  
86535                               Crime|Thriller  
86536                                  Documentary  

[86537 rows x 3 columns]

In [15]:
ratings.head(11)

userId  movieId  rating   timestamp
0        1        1     4.0  1225734739
1        1      110     4.0  1225865086
2        1      158     4.0  1225733503
3        1      260     4.5  1225735204
4        1      356     5.0  1225735119
5        1      381     3.5  1225734105
6        1      596     4.0  1225733524
7        1     1036     5.0  1225735626
8        1     1049     3.0  1225734079
9        1     1066     4.0  1225736961
10       1     1196     3.5  1225735441

In [16]:
movies.head(11)

movieId                               title  \
0         1                    Toy Story (1995)   
1         2                      Jumanji (1995)   
2         3             Grumpier Old Men (1995)   
3         4            Waiting to Exhale (1995)   
4         5  Father of the Bride Part II (1995)   
5         6                         Heat (1995)   
6         7                      Sabrina (1995)   
7         8                 Tom and Huck (1995)   
8         9                 Sudden Death (1995)   
9        10                    GoldenEye (1995)   
10       11      American President, The (1995)   

                                         genres  
0   Adventure|Animation|Children|Comedy|Fantasy  
1                    Adventure|Children|Fantasy  
2                                Comedy|Romance  
3                          Comedy|Drama|Romance  
4                                        Comedy  
5                         Action|Crime|Thriller  
6                                Comedy|Romance  
7                            Adventure|Children  
8                                        Action  
9                     Action|Adventure|Thriller  
10                         Comedy|Drama|Romance

In [18]:
from sklearn.metrics.pairwise import cosine_similarity

In [22]:
print(ratings.describe())
print(ratings.info())

             userId       movieId        rating     timestamp
count  3.383216e+07  3.383216e+07  3.383216e+07  3.383216e+07
mean   1.654380e+05  2.831348e+04  3.542540e+00  1.269362e+09
std    9.534122e+04  4.992865e+04  1.063959e+00  2.541023e+08
min    1.000000e+00  1.000000e+00  5.000000e-01  7.896520e+08
25%    8.295300e+04  1.219000e+03  3.000000e+00  1.046718e+09
50%    1.661290e+05  3.263000e+03  4.000000e+00  1.264740e+09
75%    2.474500e+05  4.049100e+04  4.000000e+00  1.496919e+09
max    3.309750e+05  2.889830e+05  5.000000e+00  1.689843e+09
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33832162 entries, 0 to 33832161
Data columns (total 4 columns):
 #   Column     Dtype  
---  ------     -----  
 0   userId     int64  
 1   movieId    int64  
 2   rating     float64
 3   timestamp  int64  
dtypes: float64(1), int64(3)
memory usage: 1.0 GB
None


In [23]:
ratings['userId'] = ratings['userId'].astype(int)
ratings['movieId'] = ratings['movieId'].astype(int)

print(ratings['userId'].max())
print(ratings['movieId'].max())

330975
288983


In [29]:
# Check the unique values of userId
unique_users = ratings['userId'].unique()
print(f"Max userId: {unique_users.max()}, Min userId: {unique_users.min()}")

# Identify problematic indices
problematic_indices = ratings[(ratings['userId'] > 1780224249) | (ratings['movieId'] > 1780224249)]
print(f"Problematic entries:\n{problematic_indices}")

Max userId: 330975, Min userId: 1
Problematic entries:
Empty DataFrame
Columns: [userId, movieId, rating, timestamp]
Index: []


In [ ]:
summary of what each part of the code does:

Loading Data: Reads the ratings data from the CSV file.

Subset Creation: Takes the first 1000 users to make the process manageable.

User-Item Matrix: Creates a matrix where rows represent users and columns represent movies. 
                   Each cell contains the rating a user gave to a movie.

Cosine Similarity: Computes the similarity between users based on their ratings using cosine similarity.

DataFrame: Converts the similarity matrix into a DataFrame for easier manipulation and analysis.

In [39]:
import pandas as pd
from sklearn.metrics.pairwise import cosine_similarity

# Load the data
ratings = pd.read_csv('ratings.csv')

# Use a manageable subset of the data
ratings_subset = ratings[ratings['userId'] < 1000]  # Taking the first 1000 users

# Create user-item matrix
user_item_matrix = ratings_subset.pivot(index='userId', columns='movieId', values='rating').fillna(0)

# Compute similarity matrix
user_similarity = cosine_similarity(user_item_matrix)
user_similarity_df = pd.DataFrame(user_similarity, index=user_item_matrix.index, columns=user_item_matrix.index)
print(user_similarity_df.head())

userId       1         2         3         4         5         6         7    \
userId                                                                         
1       1.000000  0.068407  0.054988  0.138945  0.089960  0.277858  0.082368   
2       0.068407  1.000000  0.052926  0.079995  0.153549  0.143528  0.233090   
3       0.054988  0.052926  1.000000  0.310411  0.071458  0.085309  0.012691   
4       0.138945  0.079995  0.310411  1.000000  0.115592  0.153152  0.051713   
5       0.089960  0.153549  0.071458  0.115592  1.000000  0.288597  0.144154   

userId       8         9         10   ...       990       991       992  \
userId                                ...                                 
1       0.020471  0.049785  0.209810  ...  0.037389  0.036023  0.018069   
2       0.013792  0.599754  0.060517  ...  0.023077  0.490555  0.120278   
3       0.000000  0.041481  0.191980  ...  0.000000  0.053651  0.086114   
4       0.000000  0.062696  0.172521  ...  0.000000  0.042589  0

# Define a function to get top N recommendations:

In [40]:
def get_top_n_recommendations(user_id, n=10):
    # Find similar users
    similar_users = user_similarity_df[user_id].sort_values(ascending=False).index[1:n+1]  # Skip the first user (itself)
    
    # Get the ratings of similar users
    similar_users_ratings = user_item_matrix.loc[similar_users]
    
    # Compute the weighted average of ratings
    weighted_ratings = similar_users_ratings.T.dot(user_similarity_df[user_id].sort_values(ascending=False)[1:n+1])
    
    # Filter out movies already rated by the user
    user_unrated_movies = user_item_matrix.loc[user_id][user_item_matrix.loc[user_id] == 0].index
    
    # Sort and get the top N recommendations
    recommendations = weighted_ratings.loc[user_unrated_movies].sort_values(ascending=False).head(n)
    
    return recommendations.index

# Get top 10 recommendations for user 1
recommended_movies = get_top_n_recommendations(1, 10)
print(f"Top 10 recommendations for user 1: {recommended_movies.tolist()}")


Top 10 recommendations for user 1: [318, 1198, 527, 58559, 1270, 457, 296, 3114, 8360, 6377]


# Map Movie IDs to Titles

In [42]:
# Load movie titles
movies = pd.read_csv('movies.csv')

# Get recommended movie titles
recommended_movie_titles = movies[movies['movieId'].isin(recommended_movies)]['title'].tolist()
print(f"Recommended movies for user 1: {recommended_movie_titles}")


Recommended movies for user 1: ['Pulp Fiction (1994)', 'Shawshank Redemption, The (1994)', 'Fugitive, The (1993)', "Schindler's List (1993)", 'Raiders of the Lost Ark (Indiana Jones and the Raiders of the Lost Ark) (1981)', 'Back to the Future (1985)', 'Toy Story 2 (1999)', 'Finding Nemo (2003)', 'Shrek 2 (2004)', 'Dark Knight, The (2008)']


# To evaluate how well your recommendation model is performing

In [43]:
from sklearn.model_selection import train_test_split

# Split the data
train_data, test_data = train_test_split(ratings_subset, test_size=0.2, random_state=42)

# Create user-item matrices
train_matrix = train_data.pivot(index='userId', columns='movieId', values='rating').fillna(0)
test_matrix = test_data.pivot(index='userId', columns='movieId', values='rating').fillna(0)


In [44]:
from sklearn.metrics.pairwise import cosine_similarity

# Compute similarity matrix using the training data
user_similarity_train = cosine_similarity(train_matrix)
np.fill_diagonal(user_similarity_train, 0)
user_similarity_train_df = pd.DataFrame(user_similarity_train, index=train_matrix.index, columns=train_matrix.index)


In [47]:
def predict_ratings(user_id, user_similarity_df, user_item_matrix):
    if user_id not in user_similarity_df.index:
        return pd.Series(np.zeros(user_item_matrix.shape[1]), index=user_item_matrix.columns)  # Return zero ratings if user not found
    
    similar_users = user_similarity_df[user_id].sort_values(ascending=False).index[:10]  # Top 10 similar users
    similar_users_ratings = user_item_matrix.loc[similar_users]
    weighted_ratings = similar_users_ratings.T.dot(user_similarity_df[user_id].sort_values(ascending=False)[:10])
    predicted_ratings = weighted_ratings / np.array([np.abs(user_similarity_df[user_id].sort_values(ascending=False)[:10]).sum(axis=0)])
    
    return predicted_ratings

# Predict ratings for the test users
predicted_ratings = pd.DataFrame(index=test_matrix.index, columns=test_matrix.columns)
for user_id in test_matrix.index:
    predicted_ratings.loc[user_id] = predict_ratings(user_id, user_similarity_train_df, train_matrix)


In [52]:
predicted_flat = predicted_ratings.values.flatten().astype(np.float64)
test_flat = test_matrix.values.flatten().astype(np.float64)

# Remove NaNs
mask = ~np.isnan(predicted_flat) & ~np.isnan(test_flat)
predicted_flat_clean = predicted_flat[mask]
test_flat_clean = test_flat[mask]

# Compute RMSE
rmse = sqrt(mean_squared_error(test_flat_clean, predicted_flat_clean))
print(f'RMSE: {rmse}')


RMSE: 0.37209547589225034


# the lower the RMSE, the more accurate your predictions are

# An RMSE between 0.2 and 0.5 is typically seen as quite good.

# so our model shows 0.37 which is between 0.2-0.5 